projects/pgai/pgai/alembic/configuration.py

```diff
@@ -0,0 +1,227 @@
+import re
+from dataclasses import dataclass, fields
+from datetime import timedelta
+from typing import ClassVar, Literal, Protocol, runtime_checkable
+
+from pydantic import BaseModel
+
+from pgai.vectorizer.base import (
+    BaseOllamaConfig,
+    BaseOpenAIConfig,
+    BaseProcessing,
+    BasePythonTemplate,
+    BaseVoyageAIConfig,
+    ChunkingCharacterTextSplitter,
+    ChunkingRecursiveCharacterTextSplitter,
+)
+
+
+@runtime_checkable
+class SQLArgumentProvider(Protocol):
+    def to_sql_argument(self) -> str: ...
+
+
+@runtime_checkable
+class PythonArgProvider(Protocol):
+    def to_python_arg(self) -> str: ...
+
+
+def format_sql_params(params: dict[str, str | None | bool | list[str]]) -> str:
+    """Format dictionary of parameters into SQL argument string without any quoting."""
+    formatted: list[str] = []
+    for key, value in params.items():
+        if value is None:
+            continue
+        elif isinstance(value, bool):
+            formatted.append(f"{key}=>{str(value).lower()}")
+        elif isinstance(value, list):
+            array_list = ",".join(f"E'{v}'" for v in value)
+            formatted.append(f"{key}=>ARRAY[{array_list}]")
+        elif isinstance(value, timedelta):
+            formatted.append(f"{key}=>'{value.seconds} seconds'")
+        else:
+            formatted.append(f"{key}=> '{value}'")
+    return ", ".join(formatted)
+
+
+class SQLArgumentMixin:
+    arg_type: ClassVar[str]
+    function_name: ClassVar[str | None] = None
+
+    def to_sql_argument(self) -> str:
+        # Get all fields including from parent classes
+        params = {}
+        for field_name, _field in self.model_fields.items():  # type: ignore
+            if field_name != "arg_type":
+                value = getattr(self, field_name)  # type: ignore
+                if value is not None:
+                    params[field_name] = value
+
+        if self.function_name:
+            fn_name = self.function_name
+        else:
+            base_name = self.__class__.__name__
+            # Convert camelCase to snake_case
+            base_name = re.sub("([a-z0-9])([A-Z])", r"\1_\2", base_name).lower()
+            # Remove 'config' and clean up any double underscores
+            base_name = base_name.replace("config", "").strip("_")
+            fn_name = f"{self.arg_type}_{base_name}"
+
+        return f", {self.arg_type} => ai.{fn_name}({format_sql_params(params)})"  # type: ignore
+
+
+class OpenAIConfig(BaseOpenAIConfig, SQLArgumentMixin):
```

Comment:
Naming (and I know naming discussions are always annoying) but why not stick to the sql convention we established and name this EmbeddingOpenAIConfig or EmbeddingConfigOpenAI? (and similar for others). The pro is that the name translation from sql->python is super easy and I think would be easier to understand. The con is that it's long. 

Otherwise the translation seems a bit ad-hoc. e.g. Indexing configs have "indexing" in the name but in a different spot than the sql. Let's think about this some more

Reply:
I think thats fixed with the code gen!